### PR TITLE
Fix double free and leak in RustString

### PR DESCRIPTION
**Pull request by rnestler**

### CODE REVIEW

**Comment by Minoru** [comment]
Wow, that's an excellent catch, @rnestler! How did you spot this problem?

Also, how do you feel about removing both the constructor and the operator? As of 9a25e0bbc196111be4ab48233e1caa5872113254, they are not used anywhere (I removed both, everything compiled fine, tests still pass).

**Comment by rnestler** [comment]
> How did you spot this problem?

The assignment operator caught my eye while I was looking up how to use `RustString` again. Since it looked suspicious I just wrote the tests which confirmed my suspicions. I programmed a lot of embedded C++ in the past, maybe that helped :wink: It's also one of these cases were the safety guarantees of Rust really shines :slightly_smiling_face: Especially the way moves and ownership work in Rust.

> Also, how do you feel about removing both the constructor and the operator?

Well having no code at all removes all possible bugs :wink: so I guess that would be fine. We'll just need to take care to not reintroduce bugs, if we ever need the move constructor or the move assignment.

**Comment by rnestler** [comment]
I also noticed that the tests won't rebuild if I just change `rs_utils.h`. I needed to do a `touch test/utils.cpp` before running `make tests`.

**Comment by rnestler** [comment]
I also noticed that most functions in `libnewsboat-ffi/src/utils.rs` accept raw pointers but do not handle `nullptr`. I guess this is OK since they are marked unsafe?

**Comment by rnestler** [comment]
@Minoru I removed the move constructor and move assignment and adjusted the tests.

File: test/utils.cpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 #include "3rd-party/catch.hpp"
 #include "test-helpers.h"
+#include "rs_utils.h"
```
Comment (Minoru) [inline test/utils.cpp]: Please run `make depslist` to update Make's understanding of file dependencies.

**Comment by Minoru** [comment]
> We'll just need to take care to not reintroduce bugs, if we ever need the move constructor or the move assignment.

I hope we won't need that. The whole point of that class was to hide the boilerplate of converting `char*` to `std::string` and freeing memory; I only intended it to be used as a temporary, not as a local variable or, worse yet, part of some larger structure.

> I also noticed that the tests won't rebuild if I just change rs_utils.h. I needed to do a touch test/utils.cpp before running make tests.

Yeah, we store the dependency graph in `mk/`, and have to update that directory every time we add/remove a `.cpp` file or change `#include`s. You added `#include "rs_utils.h"` without running that command, hence the behaviour you describe.

> I also noticed that most functions in libnewsboat-ffi/src/utils.rs accept raw pointers but do not handle nullptr. I guess this is OK since they are marked unsafe?

No, I just don't see much point in handling nullptr there. These functions are only used in concert with their C++ counterparts, so I can assume that e.g. `std::string::c_str()` won't give me a nullptr, and if I give the C++ side a `Box` I'll get the pointer back eventually.

I'm the one who applied `std::move` to `char*`, though, so if you have counter-arguments, please share :) I certainly don't know everything there is to know.

File: test/utils.cpp (hunk 1 above)
Comment (rnestler) [reply test/utils.cpp]: Ah that's why I needed to do `touch test/utils.cpp` when changing `rs_utils.h`. :slightly_smiling_face:

**Comment by rnestler** [comment]
> I hope we won't need that. The whole point of that class was to hide the boilerplate of converting char* to std::string and freeing memory; I only intended it to be used as a temporary, not as a local variable or, worse yet, part of some larger structure.

True. Also with the commits which actually fix the bug in the git history we'd probably notice if we'd need it back for any reason.

> Yeah, we store the dependency graph in mk/, and have to update that directory every time we add/remove a .cpp file or change #includes. You added #include "rs_utils.h" without running that command, hence the behaviour you describe.

We could run `make depslist` during CI and check if it would modify anything.

> No, I just don't see much point in handling nullptr there. These functions are only used in concert with their C++ counterparts, so I can assume that e.g. std::string::c_str() won't give me a nullptr, and if I give the C++ side a Box I'll get the pointer back eventually.

Agree, since we don't actually use raw `char` pointers this shouldn't be an issue.

**Comment by Minoru** [comment]
> We could run make depslist during CI and check if it would modify anything.

Oh, that's a good idea. Files #770 about this. I'm moving some jobs from Travis to Cirrus at the moment, so I'll add the check in the nearest future.

I'll merge this PR once the builds finish.

**Pull request merged**